activation/nipost_test.go

```diff
@@ -807,6 +807,393 @@ func TestNIPSTBuilder_PoetUnstable(t *testing.T) {
+// it properly detects added/deleted PoET services and re-registers if needed.
+func TestNIPoSTBuilder_PoETConfigChange(t *testing.T) {
+	t.Parallel()
+
+	sig, err := signing.NewEdSigner()
+	require.NoError(t, err)
+
+	challenge := types.NIPostChallenge{
+		PublishEpoch: postGenesisEpoch + 2,
+	}
+
+	const (
+		poetProverAddr  = "http://localhost:9999"
+		poetProverAddr2 = "http://localhost:9988"
+	)
+
+	t.Run("1 poet deleted BEFORE round started -> continue get proof from known poets", func(t *testing.T) {
+		db := localsql.InMemory()
+
+		require.NoError(t, nipost.AddChallenge(db, sig.NodeID(), &challenge))
+		challengeHash := wire.NIPostChallengeToWireV1(&challenge).Hash()
+
+		ctrl := gomock.NewController(t)
+		poetProver := defaultPoetServiceMock(t, ctrl, poetProverAddr)
+		poetProver.EXPECT().Proof(gomock.Any(), "1").Return(
+			&types.PoetProof{}, []types.Hash32{challengeHash}, nil,
+		)
+
+		mclock := defaultLayerClockMock(ctrl)
+
+		postClient := NewMockPostClient(ctrl)
+		nonce := types.VRFPostIndex(1)
+		postClient.EXPECT().Proof(gomock.Any(), gomock.Any()).Return(
+			&types.Post{
+				Indices: []byte{1, 2, 3},
+			}, &types.PostInfo{
+				Nonce: &nonce,
+			}, nil,
+		)
+
+		postService := NewMockpostService(ctrl)
+		postService.EXPECT().Client(sig.NodeID()).Return(postClient, nil).AnyTimes()
+
+		// successfully registered to 2 poets
+		err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+			ChallengeHash: challengeHash,
+			Address:       poetProverAddr,
+			RoundID:       "1",
+			RoundEnd:      time.Now().Add(1 * time.Second),
+		})
+		require.NoError(t, err)
+
+		err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+			ChallengeHash: challengeHash,
+			Address:       poetProverAddr2,
+			RoundID:       "1",
+			RoundEnd:      time.Now().Add(1 * time.Second),
+		})
+
+		nb, err := NewNIPostBuilder(
+			db,
+			postService,
+			zaptest.NewLogger(t),
+			PoetConfig{},
+			mclock,
+			nil,
+			WithPoetServices(poetProver), // add only 1 poet prover
+		)
+		require.NoError(t, err)
+
+		nipost, err := nb.BuildNIPost(context.Background(), sig, challengeHash, &types.NIPostChallenge{PublishEpoch: 7})
+		require.NoError(t, err)
+		require.NotNil(t, nipost)
+	})
+
+	t.Run("1 poet added BEFORE round started -> register to missing poet, get proof from both", func(t *testing.T) {
+		db := localsql.InMemory()
+
+		require.NoError(t, nipost.AddChallenge(db, sig.NodeID(), &challenge))
+		challengeHash := wire.NIPostChallengeToWireV1(&challenge).Hash()
+
+		ctrl := gomock.NewController(t)
+		poetProver := defaultPoetServiceMock(t, ctrl, poetProverAddr)
+		poetProver.EXPECT().Proof(gomock.Any(), "1").Return(
+			&types.PoetProof{}, []types.Hash32{challengeHash}, nil,
+		)
+
+		addedPoetProver := defaultPoetServiceMock(t, ctrl, poetProverAddr2)
+		addedPoetProver.EXPECT().Proof(gomock.Any(), "").Return(
+			&types.PoetProof{}, []types.Hash32{challengeHash}, nil,
+		)
```

Comment:
How about being explicit about expectations here? Right now the test doesn't verify if the builder submitted to the second poet and didn't submit to the first one, which is the essence of this test.
```suggestion
		poetProver := NewMockPoetService(ctrl)
		poetProver.EXPECT().Address().Return( poetProverAddr).AnyTimes()
		poetProver.EXPECT().Proof(gomock.Any(), "1").Return(
			&types.PoetProof{}, []types.Hash32{challengeHash}, nil,
		)

		addedPoetProver := NewMockPoetService(ctrl)
		addedPoetProver.EXPECT().
			Submit(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
			Return(&types.PoetRound{}, nil)
		addedPoetProver.EXPECT().Address().Return(poetProverAddr2).AnyTimes()
		addedPoetProver.EXPECT().Proof(gomock.Any(), "").Return(
			&types.PoetProof{}, []types.Hash32{challengeHash}, nil,
		)
```